x/forwarding/ante.go

```diff
@@ -0,0 +1,122 @@
+package forwarding
+
+import (
+	cctptypes "github.com/circlefin/noble-cctp/x/cctp/types"
+	fiattokenfactorytypes "github.com/circlefin/noble-fiattokenfactory/x/fiattokenfactory/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/noble-assets/noble/v5/x/forwarding/keeper"
+	"github.com/noble-assets/noble/v5/x/forwarding/types"
+)
+
+type Decorator struct {
+	authKeeper ante.AccountKeeper
+	keeper     *keeper.Keeper
+}
+
+var _ sdk.AnteDecorator = Decorator{}
+
+func NewAnteDecorator(keeper *keeper.Keeper, authKeeper ante.AccountKeeper) Decorator {
+	return Decorator{
+		authKeeper: authKeeper,
+		keeper:     keeper,
+	}
+}
+
+func (d Decorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	msgs := tx.GetMsgs()
+
+	err = d.CheckMessages(ctx, msgs)
+	if err != nil {
+		return ctx, err
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+func (d Decorator) CheckMessages(ctx sdk.Context, msgs []sdk.Msg) error {
+	for _, raw := range msgs {
+		if msg, ok := raw.(*authz.MsgExec); ok {
+			nestedMsgs, err := msg.GetMessages()
+			if err != nil {
+				return err
+			}
+
+			return d.CheckMessages(ctx, nestedMsgs)
+		}
+
+		switch msg := raw.(type) {
+		case *cctptypes.MsgReceiveMessage:
```

Comment:
This is an incredible idea, however on `v0.45.x`, there's sadly not much extension possible for `DeliverTx`.

How did you envision this happening?

Reply:
Something like this should work since `App` extends `BaseApp`:
```
func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
	res := app.BaseApp.DeliverTx(req)
	if res.Code != 0 {
		return res
	}

	app.Logger().Info("DeliverTx hook", "tx", req.Tx)

	for _, e := range res.Events {
		//
	}

	return res
}
```